osutil/io.go

```diff
@@ -226,3 +226,64 @@ func AtomicWriteChown(filename string, reader io.Reader, perm os.FileMode, flags
 
 	return aw.Commit()
 }
+
+// AtomicRename attempts to rename a path from oldName to newName atomically.
+func AtomicRename(oldName, newName string) error {
+	var oldDir, newDir *os.File
+
+	if !snapdUnsafeIO {
+		oldDirPath := filepath.Dir(oldName)
+		newDirPath := filepath.Dir(newName)
+
+		o, err := os.Open(oldDirPath)
+		if err != nil {
+			return err
+		}
+		oldDir = o
+		defer o.Close()
+
+		if oldDir != newDir {
+			n, err := os.Open(newDirPath)
+			if err != nil {
+				return err
+			}
+			newDir = n
+			defer n.Close()
+		}
+	}
+
+	err := os.Rename(oldName, newName)
+	if err != nil {
+		return err
+	}
+
+	if !snapdUnsafeIO {
+		var err1, err2 error
+		err1 = oldDir.Sync()
+		if newDir != nil {
+			err2 = newDir.Sync()
+		}
+		if err1 != nil {
+			return err1
+		}
+		return err2
+	}
+	return nil
+}
+
+// AtomicSymlink attempts to atomically create a symlink at linkPath, pointing
+// to a given target. The process creates a temporary symlink object pointing to
+// the target, and then proceeds to rename it atomically, replacing the
+// linkPath.
+func AtomicSymlink(target, linkPath string) error {
+
+	tmp := linkPath + "." + strutil.MakeRandomString(12) + "~"
```

Comment:
I was thinking if this is racy. Usually all schemes where you create a quasi-random name and pass it around are racy because that name can clash with someone following the same algorithm and coming up with the same name. Normally what is done is that this is coupled with a loop that has a bound number of attempts and handles failure (clash). I think this is missing here.

Reply:
I agree, this was @cmatsuoka 's suggestion as well to my other PR, so if both you and Claudio think we should do a loop here, I agree. I can push a change to this branch to do this